dev/com.ibm.ws.kernel.boot.core/src/com/ibm/ws/kernel/internal/classloader/JarFileClassLoader.java

```diff
@@ -120,22 +140,60 @@ public void close() {
 
     protected void addURL(URL url) {
         urls.add(url);
-        resourceHandlers.add(createResoureHandler(url, verify));
+        ResourceHandler handler = createResoureHandler(url, verify);
+        lock.lock();
+        try {
+            Map<String, List<ResourceHandler>> copy = new HashMap<>(handlerMap.size());
+            for (Entry<String, List<ResourceHandler>> entry : handlerMap.entrySet()) {
+                List<ResourceHandler> value = entry.getValue();
+                copy.put(entry.getKey(), value.size() == 1 ? value : new ArrayList<>(value));
+            }
+            populateHandlerMap(copy, handler);
+            handlerMap = copy;
+        } finally {
+            lock.unlock();
+        }
+        resourceHandlers.add(handler);
     }
 
     public URL[] getURLs() {
         return urls.toArray(new URL[0]);
     }
 
+    protected Class<?> _loadClass(String name) throws ClassNotFoundException {
+        Class<?> result = null;
+
+        int index = name.lastIndexOf('.');
+        String pkg = index < 0 ? "" : name.substring(0, index);
+
+        List<ResourceHandler> pkgHandlers = handlerMap.get(pkg);
+
+        if (pkgHandlers != null) {
```

Comment:
And there's the crux of the solution.  I had trouble locating FAT classes that directly exercise `JarURLClassloader.addUrl()`.  Please ensure the solution works when URLs are added dynamically to either the `BootstrapChildFirstJarFileClasloader` or its base class, `JarFileClassloader`.

Reply:
`addURL()` is called by the OSGi code.  I tried removing `addURL()` to not need to handle the dynamic portion, but it is necessary.   With the logic that I already added in `addURL()`, it does work now.  Without adding the logic to that method to call the `populateHandlerMap`, no features would start.  Only the kernel would start.

Reply:
Verified to work -- thank you.